Core/include/Acts/Vertexing/SeedVertexFinder.ipp

```diff
@@ -0,0 +1,656 @@
+    return false;
+  }
+
+  // norm to z-axis and to the ray
+  Acts::Vector3 norm{-1. * direction[1], 1. * direction[0], 0};
+  // nearest distance from the ray to z-axis
+  Acts::ActsScalar dist =
+      std::fabs(start_point.dot(norm)) / std::sqrt(norm.dot(norm));
+  if (dist > m_cfg.maxRPosition) {
+    return false;
+  }
+
+  // cross product of direction and norm
+  Acts::Vector3 direction_x_n = direction.cross(norm);
+  // z coordinate of the nearest distance from the ray to z-axis
+  Acts::ActsScalar zdist = direction_x_n.dot(start_point) / (norm.dot(norm));
+  if (std::fabs(zdist) > m_cfg.maxZPosition) {
+    return false;
+  }
+
+  if (m_cfg.minimalizeWRT == "rays") {
+    // save for later
+    triplet.ray = ray;
+  }
+
+  return true;
+}
+
+template <typename spacepoint_t>
+std::pair<Acts::Vector3, Acts::ActsScalar>
+Acts::SeedVertexFinder<spacepoint_t>::makePlaneFromTriplet(
+    const Acts::SeedVertexFinder<spacepoint_t>::Triplet triplet) const {
+  Acts::Vector3 a{triplet.a.x(), triplet.a.y(), triplet.a.z()};
+  Acts::Vector3 b{triplet.b.x(), triplet.b.y(), triplet.b.z()};
+  Acts::Vector3 c{triplet.c.x(), triplet.c.y(), triplet.c.z()};
+
+  Acts::Vector3 ba = b - a, ca = c - a;
+
+  Acts::Vector3 abg = ba.cross(ca);
+  // vector (alpha,beta,gamma) normalized to unity for convenience
+  abg /= std::sqrt(abg.dot(abg));
+  Acts::ActsScalar delta = -1. * abg.dot(a);
+
+  // plane (alpha*x + beta*y + gamma*z + delta = 0), splitted to {{alpha, beta,
+  // gamma}, delta} for convenience
+  return {abg, delta};
+}
+
+template <typename spacepoint_t>
+Acts::Vector3 Acts::SeedVertexFinder<spacepoint_t>::findClosestPointFromPlanes(
+    const std::vector<Acts::SeedVertexFinder<spacepoint_t>::Triplet>& triplets)
+    const {
+  // define function f = sum over all triplets [distance from an unknown point
+  // (x_0,y_0,z_0) to the plane defined by the triplet] find minimum of "f" by
+  // partial derivations over x_0, y_0, and z_0 each derivation has parts
+  // lineary depending on x_0, y_0, and z_0 (will fill A[deriv][3]) or to
+  // nothing (will fill B[deriv]) solve A*(x_0,y_0,z_0) = B
+
+  Acts::Vector3 vtx = Acts::Vector3::Zero();
+  Acts::Vector3 vtx_prev{m_cfg.rMaxFar, m_cfg.rMaxFar, m_cfg.maxAbsZ};
+
+  // (alpha-beta-gamma, delta), distance
+  std::vector<
+      std::pair<std::pair<Acts::Vector3, Acts::ActsScalar>, Acts::ActsScalar>>
+      triplets_with_planes;
+  triplets_with_planes.reserve(triplets.size());
+
+  for (const auto& triplet : triplets) {
+    auto abgd = makePlaneFromTriplet(triplet);
+    triplets_with_planes.emplace_back(abgd, -1.);
+  }
+
+  // elements of the linear equations to solve
+  Acts::SymMatrix3 A = Acts::SymMatrix3::Zero();
+  Acts::Vector3 B = Acts::Vector3::Zero();
+  for (const auto& triplet : triplets_with_planes) {
+    const auto& abg = triplet.first.first;
+    const auto& delta = triplet.first.second;
+
+    A += 2. * (abg * abg.transpose());
```

Comment:
is `abg` a vector? if so consider using `squaredNorm`

Reply:
Hi @andiwand, yes, `abg` is a vector but the desired output is a matrix. I believe `squaredNorm` would return a scalar.

Cheers, Petr